core/src/cluster_info.rs

```diff
@@ -2272,95 +2276,114 @@ impl ClusterInfo {
-        let len = crds_values.len();
-
-        let shred_version = self
-            .lookup_contact_info(from, |ci| ci.shred_version)
-            .unwrap_or(0);
-        Self::filter_by_shred_version(
-            from,
-            &mut crds_values,
-            shred_version,
-            self.my_shred_version(),
-        );
-        let filtered_len = crds_values.len();
+        self.stats
+            .push_message_count
+            .add_relaxed(messages.len() as u64);
+        // Obtain shred versions of the origins.
+        let shred_versions: Vec<_> = {
+            let gossip = self.gossip.read().unwrap();
+            messages
+                .iter()
+                .map(|(from, _)| match gossip.crds.get_contact_info(from) {
+                    None => 0,
+                    Some(info) => info.shred_version,
+                })
+                .collect()
+        };
+        // Filter out data if the origin has different shred version.
+        let self_shred_version = self.my_shred_version();
+        let num_crds_values: u64 = messages.iter().map(|(_, data)| data.len() as u64).sum();
+        let messages: Vec<_> = messages
+            .into_iter()
+            .zip(shred_versions)
+            .filter_map(|((from, mut crds_values), shred_version)| {
+                Self::filter_by_shred_version(
+                    &from,
+                    &mut crds_values,
+                    shred_version,
+                    self_shred_version,
+                );
+                if crds_values.is_empty() {
+                    None
+                } else {
+                    Some((from, crds_values))
+                }
+            })
+            .collect();
+        let num_filtered_crds_values = messages.iter().map(|(_, data)| data.len() as u64).sum();
         self.stats
             .push_message_value_count
-            .add_relaxed(filtered_len as u64);
+            .add_relaxed(num_filtered_crds_values);
         self.stats
             .skip_push_message_shred_version
-            .add_relaxed((len - filtered_len) as u64);
-
-        let updated: Vec<_> = self
-            .time_gossip_write_lock("process_push", &self.stats.process_push_message)
-            .process_push_message(from, crds_values, timestamp());
-
-        let updated_labels: Vec<_> = updated.into_iter().map(|u| u.value.label()).collect();
-        let prunes_map: HashMap<Pubkey, HashSet<Pubkey>> = self
+            .add_relaxed(num_crds_values - num_filtered_crds_values);
+        // Update crds values and obtain updated keys.
+        let updated_labels: Vec<_> = {
+            let mut gossip =
+                self.time_gossip_write_lock("process_push", &self.stats.process_push_message);
+            let now = timestamp();
+            messages
+                .into_iter()
+                .flat_map(|(from, crds_values)| {
+                    gossip.process_push_message(&from, crds_values, now)
```

Comment:
Out of curiosity, how expensive is this/rest of this function compared to the prunes that we were able to but in a thread pool below (i.e. what percentage of the `solana-listen` thread's compute? 

For future reference, if we ever want to break up these HashMaps, I've been using `DashMap` in the accounts changes which shards a HashMap for more concurrency: https://github.com/solana-labs/solana/pull/12126/files#diff-6431a3eec6e102bbcffb94387cda89efbe02002649745273fb575234c0018de4R17

Reply:
From the runtime profiles, the majority time of this function is spent making the prune messages. There is a significant overhead with rayon parallel iterators, which is justified if the processing takes enough time. So, for the other pieces of this function, I would rather not use thread-pool for now.